### PR TITLE
Add RandomGenerator class that unifies randomness across backends.

### PR DESCRIPTION
**Pull request by ehpor**
Related to #299 but independent.

This PR adds a `RandomGenerator` class. The `RandomGenerator` class provides a unified and consistent API for generating random numbers across various array API namespaces such as NumPy, CuPy, PyTorch, and JAX. Its primary purpose is to abstract away the specific implementations and nuances of random number generation in these different libraries, allowing users to write generic code that works regardless of the underlying array backend. It initializes an appropriate random number generator based on the provided namespace and an optional seed.

The class exposes common methods for sampling from statistical distributions, including `normal()` (Gaussian), `poisson()`, and `gamma()`. It internally manages the state of the random number generator, handling differences such as explicit PRNG key splitting for JAX's functional paradigm or state management for PyTorch. Additionally, it includes a `copy()` method to create an independent duplicate of the current random state.

I'm also sneaking in a Ruff config change to ignore python 3.8 warnings (which is a Python version that is no longer supported). It was annoying me, sorry.

Also, since most backends are not tested by CI, the coverage will go down despite my best efforts to add comprehensive tests. Tests will be performed if the library is installed.

### CODE REVIEW

File: tests/test_math.py [hunk 1]
```diff
@@ -55,3 +56,78 @@ def test_fft_acceleration(func, method, dtype_in, dtype_out):
 
         assert np.allclose(y_numpy, y_method, rtol=rtol, atol=rtol)
         assert y_method.dtype == dtype_out
+
+
+def test_random_generator_distribution(xp):
+    # Get samples from the Standard Normal distribution.
+    rng = RandomGenerator(xp, seed=42)
+    samples = rng.normal(size=(10000,))
```
Comment (MatthijsMars) [inline tests/test_math.py:64]: Should probably test this on non-default mean and standard deviation to show that also mean and standard deviation are propagated properly

File: tests/test_math.py [hunk 2]
```diff
@@ -55,3 +56,78 @@ def test_fft_acceleration(func, method, dtype_in, dtype_out):
...
+def test_random_generator_gamma(xp):
+    # Get samples from the Gamma distribution.
+    rng = RandomGenerator(xp, seed=42)
+    samples = rng.gamma(scale=2.0, shape_param=2.0, size=(10000,))
+
+    # Basic statistical tests (mean should be shape * scale = 2 * 2 = 4)
+    assert np.isclose(float(xp.mean(samples)), 4.0, atol=0.5)
+
+
+def test_random_generator_reproducible(xp):
```
Comment (MatthijsMars) [inline tests/test_math.py:89]: I think the bottom few tests could be parametrised for the different functions to make sure gamma and poisson also follow this behaviour.

**Review by MatthijsMars — COMMENTED** [review]
And yes I think we should most definitely add jax and pytorch to the test environment. Easiest would probably be to just have two environments, one for windows and one for unix. Or we could probably add a step in the CI that checks if it is not windows and then install pytorch and jax. Then all the code (or at least what has currently been adapted to the multibackend) should be tested in Jax, cupy and pytorch if installed in the environment.

File: tests/test_math.py (hunk 1 above)
Comment (ehpor) [reply tests/test_math.py:64]: Great point. I'll change that. Same for the other two distributions.

File: hcipy/_math/random.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        array
+            An array of specified shape filled with random samples from the
+            normal distribution.
+        """
+        if size is None:
+            size = (1,)
+        elif not isinstance(size, tuple):
+            size = (size,)
+
+        if is_numpy_namespace(self.xp) or is_cupy_namespace(self.xp):
```
Comment (MatthijsMars) [inline hcipy/_math/random.py]: Is it necessary to do this if-else every time we generate a new random number? The namespace is inferred only at __init__ so we can also do all the if-else in the init or some other function that constructs the corresponding function links or lambda functions if necessary. 

(same for all other distributions)

**Review by MatthijsMars — COMMENTED** [review]
I checked the CI and it now did install the optional dependencies (like torch and jax) and the conftest showed that cupy wasn't installed on one of them so that also works. Are the random generator tests in the CI now also actually running for each of the different backends? If not, should we include this? I think turning on multiple backend testing for the whole codebase is right now not a great option but we should automatically test the parts of the code that are compatible and implemented now already.

File: hcipy/_math/random.py (hunk 3 above)
Comment (ehpor) [reply hcipy/_math/random.py]: I don't think that would speed things up much. These checks are ~50ns each. But I'll do some benchmarks of the whole implementation:

```python
rng1 = make_random_generator(np)  # new
rng2 = RandomGenerator(np)        # old

print('normal')
%timeit rng1.normal(size=(1,))
%timeit rng2.normal(size=(1,))
%timeit rng1.normal(size=(100,))
%timeit rng2.normal(size=(100,))
%timeit rng1.normal(size=(100,100))
%timeit rng2.normal(size=(100,100))
print('poisson')
%timeit rng1.poisson(size=(1,))
%timeit rng2.poisson(size=(1,))
%timeit rng1.poisson(size=(100,))
%timeit rng2.poisson(size=(100,))
%timeit rng1.poisson(size=(100,100))
%timeit rng2.poisson(size=(100,100))
print('gamma')
%timeit rng1.gamma(size=(1,))
%timeit rng2.gamma(size=(1,))
%timeit rng1.gamma(size=(100,))
%timeit rng2.gamma(size=(100,))
%timeit rng1.gamma(size=(100,100))
%timeit rng2.gamma(size=(100,100))
```
yields
```
normal
621 ns ± 1.83 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
662 ns ± 2.17 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
1.09 μs ± 3.91 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
1.11 μs ± 7.05 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
47.8 μs ± 877 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
47.4 μs ± 76.5 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
poisson
536 ns ± 1.65 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
563 ns ± 2.72 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
2.99 μs ± 25.3 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
3.02 μs ± 11.8 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
249 μs ± 3.71 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
248 μs ± 1.33 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
gamma
631 ns ± 3.84 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
671 ns ± 9.8 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
1.11 μs ± 7.71 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
1.15 μs ± 6.5 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
49.2 μs ± 203 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
49.4 μs ± 54.4 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

So 30-50ns reduction per function call, at the cost of 60 extra lines of code.

**Comment by ehpor** [comment]
@MatthijsMars yes, the tests for the RandomGenerator class are running for each backend that is installed. This is done using a pytest fixture `xp`. Backends that are not installed are skipped with a message indicating that they are not installed. You can check this in the code coverage (see the first comment, or [this direct link](https://app.codecov.io/gh/ehpor/hcipy/pull/301?dropdown=coverage&src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Emiel+Por)).

File: hcipy/_math/random.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+
+        self._rng = self._xp.random.default_rng(seed)
+
+    def copy(self):
+        res = super().copy()
+        res._rng = copy.deepcopy(self._rng)
+
+        return res
+
+    def normal(self, mean=0.0, std=1.0, *, size=None):
```
Comment (MatthijsMars) [inline hcipy/_math/random.py:194]: As mentioned, I think there should be docstrings here and for all other functions.

**Review by MatthijsMars — COMMENTED** [review]
@ehpor Implementation and testing now look good to me. Only thing I would add is copying the copy, normal, gamma, etc. function docstrings to all the dedicated backend classes so each derived class has the proper documentation. Then random generator objects for a specific backend will have proper documentation

**Comment by ehpor** [comment]
@MatthijsMars docstrings for functions inherit from the parent class (since Python 3.5).

**Comment by MatthijsMars** [comment]
@ehpor My bad. Then it looks good to me. I'll approve it for merging

**Review by MatthijsMars — DISMISSED** [review]

**Comment by ehpor** [comment]
@MatthijsMars I accidentally did a merge update rather than a rebase update (stupid phone buttons). Fixed now. Can you reapprove?

**Review by MatthijsMars — APPROVED** [review]
